server/src/main/java/io/crate/replication/logical/plan/DropSubscriptionPlan.java

```diff
@@ -51,8 +65,61 @@ public void executeOrFail(DependencyCarrier dependencies,
                               RowConsumer consumer,
                               Row params,
                               SubQueryResults subQueryResults) throws Exception {
-        var request = new DropSubscriptionRequest(analyzedDropSubscription.name(), analyzedDropSubscription.ifExists());
-        dependencies.dropSubscriptionAction()
-            .execute(request, new OneRowActionListener<>(consumer, rCount -> new Row1(rCount == null ? -1 : 1L)));
+
+        List<RelationName> tables = InformationSchemaIterables.tablesStream(dependencies.schemas())
+            .filter(t -> t instanceof DocTableInfo)
+            .filter(t -> analyzedDropSubscription.name().equals(REPLICATION_SUBSCRIPTION_NAME.get(t.parameters())))
+            .map(t -> t.ident())
+            .collect(Collectors.toList());
+
+        final List<ChainableAction<Long>> actions = new ArrayList<>();
+
+        // Step 1 - Close subscribed tables and consequently stop tracking and remove retention lease.
+        actions.add(new ChainableAction<>(
+            () -> dependencies.alterTableOperation()
+                .executeAlterTableOpenClose(tables, false, null)
+                .exceptionally(error -> {
+                    throw new IllegalStateException(
+                        "Couldn't close subscribed tables, please retry DROP SUBSCRIPTION command."
+                            + error.getMessage(), error
+                    );
+                }),
+            () -> CompletableFuture.completedFuture(-1L)
+        ));
+
+        // Step 2
+        // Drop setting and subscription
+        actions.add(new ChainableAction<>(
+            () -> {
+                FutureActionListener<AcknowledgedResponse, Long> listener = new FutureActionListener<>(r -> 1L);
+                var request = new DropSubscriptionRequest(analyzedDropSubscription.name(), analyzedDropSubscription.ifExists());
+                dependencies.dropSubscriptionAction().execute(request, listener);
+                return listener.exceptionally(error -> {
+                    throw new IllegalStateException(
+                        "Couldn't update metadata, please retry DROP SUBSCRIPTION command."
+                            + error.getMessage(), error
+                    );
+                });
+            },
+            () -> CompletableFuture.completedFuture(-1L)
+        ));
+
+
+        // Step 3
+        // Reopen table to update table engine to normal (Operation will be reset back to ALL after step 2)
+        actions.add(new ChainableAction<>(
+            () -> dependencies.alterTableOperation()
+                .executeAlterTableOpenClose(tables, true, null)
+                .exceptionally(error -> {
+                    throw new IllegalStateException(
+                        "Couldn't reopen tables, please run command ALTER TABLE OPEN for all subscribed tables."
```

Comment:
Ca we maybe include the  tables inside the error message?

Reply:
Should I add some check for tables size? So that it't not some unreadable long text if it's FOR ALL TABLES and number of tables is big.

Reply:
good point 👍